src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs

```diff
@@ -185,13 +186,35 @@ private static Task<MSB.Execution.BuildResult> BuildAsync(MSB.Execution.BuildMan
                     {
                         buildManager.CancelAllSubmissions();
                         buildManager.EndBuild();
-                        registration.Dispose();
                     }
                     finally
                     {
                         taskSource.TrySetCanceled();
                     }
                 });
+
+                // Dispose of the registration as soon as we no longer need it
+                var _ = DisposeWhenCompleteAsync(taskSource.Task, registration);
+
+                async Task DisposeWhenCompleteAsync<T>(Task task, T disposable)
```

Comment:
can you move the local function to the bottom of the method (or as a sibling method).  I find it *very* hard to read methods that have a local function randomly interspersed with the code.